### ISSUE TITLE
Implement pi and n heterozygosity metrics

### ISSUE THREAD

**Issue by niemasd** [ucsd-ccbb/covid_sequencing_analysis_pipeline#104]
To compute the "pi" metric, you would use this program (the compile command is simple and is in a code comment on Line 3):

https://github.com/Niema-Docker/pi_from_pileup/blob/main/pi_from_pileup.cpp

You would run it as follows:

```bash
pi_from_pileup PILEUP_FILE MIN_DEPTH | tail -1 | cut -f3
```

I added `tail -1 | cut -f3` because my `pi_from_pileup` code computes the diversity at every position of the genome, but the pi metric is the third element in the last line. The `MIN_DEPTH` argument is optional, and if you don't provide it (i.e., you just run `pi_from_pileup PILEUP_FILE`), it'll use the default value of 10.

To compute Karthik's "n" metric, which was defined as "number of positions with alternate allele frequency between 0.05 and 0.95", you would do the following:

```bash
cat VARIANTS_TSV | cut -f11 | grep -v "^ALT_FREQ" | numlist -gte0.05 | numlist -lte0.95 | wc -l
```

This depends on my `numlist` program, which you can find here:

https://github.com/niemasd/tools/blob/master/numlist.cpp

The `Makefile` is in the same GitHub repo, but the compilation command is simple:

```bash
g++ -Wall -pedantic -O3 -std=c++11 -o numlist numlist.cpp
```

**Comment by AmandaBirmingham** [comment]
@kmfisch just to make sure I understand: each of these metrics is calculated at the level of an individual sample?

**Comment by kmfisch** [comment]
Yes and we'll get code from Niema and Karthik to implement.

**Comment by niemasd** [comment]
@kmfisch @AmandaBirmingham

# pi
Here's the code to compute pi from a samtools pile-up file:

https://github.com/Niema-Docker/pi_from_pileup/blob/main/pi_from_pileup.cpp

Here's how you'd run it:

```bash
pi_from_pileup PILEUP_FILE MIN_DEPTH
```

The `MIN_DEPTH` argument is optional, and if you don't provide it (i.e., you just run `pi_from_pileup PILEUP_FILE`), it'll use the default value of 10.

The program actually prints the diversity at every single position, and *then* prints the pi statistic at the end, so to just get the pi statistic, you could do the following:

```bash
pi_from_pileup PILEUP_FILE | tail -1 | cut -f3
```

# Karthik's *n* metric
This was defined as "number of positions with alternate allele frequency between 0.05 and 0.95". We can wait for his code as well, but the following should compute it equivalently:

```bash
cat VARIANTS_TSV | cut -f11 | grep -v "^ALT_FREQ" | numlist -gte0.05 | numlist -lte0.95 | wc -l
```

This depends on my `numlist` program, which you can find here:

https://github.com/niemasd/tools/blob/master/numlist.cpp

The `Makefile` is in the same GitHub repo, but the compilation command is simple:

```bash
g++ -Wall -pedantic -O3 -std=c++11 -o numlist numlist.cpp
```

**Comment by niemasd** [comment]
To follow up from the discrepancy between my *n* calculations and Karthik's, from his email, we should be able to use the following command to calculate *n* instead:

```bash
cat VARIANT_TSV | awk '$11 != "ALT_FREQ" && $11 >= 0.05 && $11 <= 0.95' | grep -v "ALT_FREQ" | cut -f 2,3,4,11 | sort | uniq | wc -l
```

I'll test and update

**Comment by niemasd** [comment]
Following up again, my command counts each position only once, but it seems as though a single position may have multiple different mutations, which we *do* want to count multiple times (we just don't want to count the *same* mutation at the *same* position as it impacts different *ORFs*). Thus, instead of just looking at unique values in column 2 (genome position) of the variants TSV after filtering, we would need to look at columns 2 and 4 (genome position + alt allele). Here is the updated command:

```bash
cat VARIANT_TSV | awk '$11 != "ALT_FREQ" && $11 >= 0.05 && $11 <= 0.95' | grep -v "ALT_FREQ" | cut -f2,4 | sort | uniq | wc -l
```

I'm rerunning and will send to @AmandaBirmingham to verify it matches what Karthik sent before

**Comment by rob-knight** [comment]
Closed as duplicate of #104.
